compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/StaticLambdaLowering.kt

```diff
@@ -0,0 +1,110 @@
+import org.jetbrains.kotlin.ir.IrElement
+import org.jetbrains.kotlin.ir.IrStatement
+import org.jetbrains.kotlin.ir.builders.irCall
+import org.jetbrains.kotlin.ir.builders.irExprBody
+import org.jetbrains.kotlin.ir.declarations.IrClass
+import org.jetbrains.kotlin.ir.declarations.IrFile
+import org.jetbrains.kotlin.ir.declarations.IrSimpleFunction
+import org.jetbrains.kotlin.ir.expressions.*
+import org.jetbrains.kotlin.ir.expressions.impl.IrGetFieldImpl
+import org.jetbrains.kotlin.ir.symbols.IrTypeParameterSymbol
+import org.jetbrains.kotlin.ir.types.IrType
+import org.jetbrains.kotlin.ir.types.classifierOrNull
+import org.jetbrains.kotlin.ir.util.constructedClass
+import org.jetbrains.kotlin.ir.util.primaryConstructor
+import org.jetbrains.kotlin.ir.visitors.IrElementTransformerVoid
+import org.jetbrains.kotlin.ir.visitors.IrElementVisitorVoid
+import org.jetbrains.kotlin.ir.visitors.acceptChildrenVoid
+import org.jetbrains.kotlin.utils.addToStdlib.safeAs
+
+internal val staticLambdaPhase = makeIrFilePhase(
+    ::StaticLambdaLowering,
+    name = "StaticLambdaPhase",
+    description = "Turn static callable references into singletons"
+)
+
+class StaticLambdaLowering(val backendContext: JvmBackendContext) : FileLoweringPass, IrElementTransformerVoid() {
+    override fun lower(irFile: IrFile) = irFile.transformChildrenVoid()
+
+    override fun visitClass(declaration: IrClass): IrStatement {
+        declaration.transformChildrenVoid()
+        if (declaration.isSyntheticSingleton) {
+            declaration.declarations += backendContext.declarationFactory.getFieldForObjectInstance(declaration).also { field ->
+                field.initializer = backendContext.createIrBuilder(field.symbol).run {
+                    irExprBody(irCall(declaration.primaryConstructor!!))
+                }
+            }
+        }
+        return declaration
+    }
+
+    override fun visitConstructorCall(expression: IrConstructorCall): IrExpression {
+        val constructor = expression.symbol.owner
+        if (!constructor.constructedClass.isSyntheticSingleton)
+            return super.visitConstructorCall(expression)
+
+        val instanceField = backendContext.declarationFactory.getFieldForObjectInstance(constructor.constructedClass)
+        return IrGetFieldImpl(expression.startOffset, expression.endOffset, instanceField.symbol, expression.type)
+    }
+
+    // Recognize callable references with no value or type arguments. The only type arguments in Kotlin stem from usages of
+    // reified type parameters, which we unfortunately don't record as parameters so we have to check the body of the class.
+    private val IrClass.isSyntheticSingleton: Boolean
+        get() = (origin == JvmLoweredDeclarationOrigin.LAMBDA_IMPL || origin == JvmLoweredDeclarationOrigin.FUNCTION_REFERENCE_IMPL)
+                && primaryConstructor!!.valueParameters.isEmpty()
+                && !containsReifiedTypeParameters
```

Comment:
I'd probably add some tests where this condition (no reified type parameters used) is relevant.

Reply:
There are already blackbox tests which test all of these conditions.